src/lib.rs

```diff
@@ -271,21 +275,67 @@ impl<T> Sender<T> {
         // Don't run our Drop implementation if send was called, any cleanup now happens here
         mem::forget(self);
 
+        // SAFETY: The channel exists on the heap for the entire duration of this method and we
+        // only ever acquire shared references to it. Note that if the receiver disconnects it
+        // does not free the channel.
         let channel = unsafe { channel_ptr.as_ref() };
 
         // Write the message into the channel on the heap.
+        // SAFETY: The receiver only ever accesses this memory location if we are in the MESSAGE
+        // state, and since we're responsible for setting that state, we can guarantee that we have
+        // exclusive access to this memory location to perform this write.
         unsafe { channel.write_message(message) };
 
         // Set the state to signal there is a message on the channel.
-        match channel.state.swap(MESSAGE, SeqCst) {
+        // ORDERING: we use release ordering to ensure the write of the message is visible to the
+        // receiving thread. The EMPTY and DISCONNECTED branches do not observe any shared state,
+        // and thus we do not need acquire orderng. The RECEIVING branch manages synchronization
+        // independent of this operation.
+        //
+        // EMPTY + 1 = MESSAGE
+        // RECEIVING + 1 = UNPARKING
+        // DISCONNECTED + 1 = invalid, however this state is never observed
+        match channel.state.fetch_add(1, Release) {
             // The receiver is alive and has not started waiting. Send done.
             EMPTY => Ok(()),
             // The receiver is waiting. Wake it up so it can return the message.
+            #[allow(unreachable_code)]
```

Comment:
I'd prefer another way to work around this issue. My understanding is that this is only a problem when no features are enabled, right? Instead of these blanked-allows that risks hiding future issues I think we should employ someting like this: https://github.com/faern/oneshot/commit/0d3e79175d6b217146040a58ae8c2cdbeb0c12eb

Feel free to cherry-pick this commit into the PR in order to get rid of these `#[allow..]`s

Reply:
The problem is that with all features disabled, the warnings became incredibly infectious, bubbling all the way up to where even more imports had to be gated (and that's already messy), and one of the states being unused (`UNPARKING`). I can propagate all those up if you want me too but the end result might not be much better than this already non-ideal solution.

Reply:
I'm not able to reproduce that. If I cherry-pick my linked commit onto this PR branch and build/test with no features I get no warnings. Neither with nor without loom.

Reply:
Oh I glanced at that commit too quickly and didnt see what it was doing. That seems like a reasonable alternative.